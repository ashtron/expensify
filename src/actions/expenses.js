import uuid from 'uuid';
import database from '../firebase/firebase';

// component calls action generator
// action generator returns the object (function for async actions)
// component dispatches the object (function)
// redux store changes (function runs: has ability to dispatch other actions,
// do other stuff too (like address firebase))

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "", 
            note = "", 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    console.log("fetching data from firebase");
    return (dispatch) => {
        // fetch expenses from firebase, parse into array
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            // dispatch SET_EXPENSES
            dispatch(setExpenses(expenses));
        });
    };
};