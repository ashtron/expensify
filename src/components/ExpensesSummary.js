import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
    const filteredExpensesCount = props.unfilteredExpensesCount-props.filteredExpensesCount;
    const filteredExpenseWord = (props.filteredExpensesCount === 1 ) ? 'expense' : 'expenses';
    const unfilteredExpenseWord = (filteredExpensesCount === 1 ) ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(props.filteredExpensesTotal / 100).format('$0,0.00');

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{props.filteredExpensesCount}</span> {filteredExpenseWord} totalling <span>{ formattedExpensesTotal }</span>.</h1>
                <h3 className="page-header__title">Filtering <span>{filteredExpensesCount}</span> {unfilteredExpenseWord} from view.</h3>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const filteredExpenses = selectExpenses(state.expenses, state.filters);
    return {
        unfilteredExpensesCount: state.expenses.length,
        filteredExpensesCount: filteredExpenses.length,
        filteredExpensesTotal: selectExpensesTotal(filteredExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);