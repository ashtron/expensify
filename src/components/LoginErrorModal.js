import React from 'react';
// import { connect } from 'react-redux';
// import { hideModal } from '../actions/modals';
import Modal from 'react-modal';

Modal.setAppElement("#app");

const LoginError = (props) => {
    return (
        <Modal
            isOpen={!!props.showModal}
            onRequestClose={props.onRequestClose}
            contentLabel="Login error"
            closeTimeoutMS={200}
            className="modal"
        >
            <h3 className="modal__title">Error code {}</h3>
        </Modal>
    );
};

export default LoginError;

// export default connect(null, { hideModal })(Confirmation);