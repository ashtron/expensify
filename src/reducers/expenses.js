const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // weird syntax: returns the concatenation of the old array with action.expense
            return [
                ...state, action.expense
            ];
            // return state.concat(action.expense);
        case 'REMOVE_EXPENSE':
            // return new array with expenses after filtering out the
            // expense with the given id from the previous array
            // if the function returns true, the item is kept
            // if the function returns false, the item is removed
            // so we check if the id is NOT the same, therefore
            // we allow all the ones whose IDs are different than the given
            // id to stay in the array, and the one whose ID is the same
            // gets filtered out
            // we use object destructuring here to get id, the first
            // argument in the expense object array that was passed in
            // so basically there are imaginary commas after it
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                // only edit the expense if it's the right object
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates // overrides the conflicting ones
                    };
                } else
                    return expense; // do nothing
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};