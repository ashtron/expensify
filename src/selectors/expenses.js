import moment from 'moment';

// get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        // show most recent first (higher createdAt value)
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        // show most expensive first (higher amount value)
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};