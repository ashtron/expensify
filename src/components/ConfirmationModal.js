import React from 'react';
import { connect } from 'react-redux';
import { hideModal } from '../actions/modals';
import Modal from 'react-modal';

Modal.setAppElement("#app");

const Confirmation = (props) => {
    return (
        <Modal
            isOpen={!!props.showModal}
            onRequestClose={props.onRequestClose}
            contentLabel="Confirm removal"
            closeTimeoutMS={200}
            className="modal"
        >
            <h3 className="modal__title">Are you sure?</h3>
            <button onClick={props.onSubmit}>
                Yes
            </button>
            <button onClick={props.onRequestClose}>
                No
            </button>
        </Modal>
    );
};

export default Confirmation;

// export default connect(null, { hideModal })(Confirmation);