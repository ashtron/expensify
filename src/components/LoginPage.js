import React from 'react';
import { connect } from 'react-redux';
import { startLoginWithGoogle, startLoginWithTwitter, startLoginWithGithub } from '../actions/auth';
import LoginErrorModal from './LoginErrorModal';

export const LoginPage = ({ startLoginWithGoogle, startLoginWithTwitter, startLoginWithGithub }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expens1fy</h1>
            <p>It's time to get your expenses under control.</p>
            <button className="button button--login" onClick={startLoginWithGoogle}>Log in with Google</button>
            <button style={{background: "#464b5e"}} className="button button--login" onClick={startLoginWithTwitter}>Log in with Twitter</button>
            <button style={{background: "#364051"}} className="button button--login" onClick={startLoginWithGithub}>Log in with GitHub</button>
        </div>
        {/* <LoginErrorModal 
            showModal={}
            onRequestClose={this.toggleModal}
        /> */}
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLoginWithGoogle: () => dispatch(startLoginWithGoogle()),
    startLoginWithTwitter: () => dispatch(startLoginWithTwitter()),
    startLoginWithGithub: () => dispatch(startLoginWithGithub())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);