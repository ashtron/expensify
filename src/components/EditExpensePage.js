import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import ConfirmationModal from './ConfirmationModal';

export class EditExpensePage extends React.Component {

    constructor(props) {
        super(props);
      
        this.state = {
          showModal: false
        };
    }

    toggleModal = () => {
        this.setState({
            showModal: !this.state.showModal
        });
    }
    
    onEditSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemoveSubmit = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.toggleModal();
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        onSubmit={this.onEditSubmit}
                        expense={this.props.expense}
                    />
                    <button
                        className="button button--secondary"
                        onClick={this.toggleModal}
                    >Remove expense</button>
                    <ConfirmationModal 
                        showModal={this.state.showModal}
                        onRequestClose={this.toggleModal}
                        onSubmit={this.onRemoveSubmit}
                        // ariaHideApp={false}
                    />
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);