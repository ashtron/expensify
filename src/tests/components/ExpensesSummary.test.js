import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import selectExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

test("should render total for two expenses and pluralize correctly", () => {
    const wrapper = shallow(<ExpensesSummary
        expensesCount={2}
        expensesTotal={selectExpensesTotal([expenses[0], expenses[1]])}
    />);
    expect(wrapper).toMatchSnapshot();
});

test("should render total for one expense and pluralize correctly", () => {
    const wrapper = shallow(<ExpensesSummary
        expensesCount={1}
        expensesTotal={selectExpensesTotal([expenses[0]])}
    />);
    expect(wrapper).toMatchSnapshot();
})