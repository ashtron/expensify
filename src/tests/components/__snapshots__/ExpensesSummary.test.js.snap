// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render total for one expense and pluralize correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
  </p>
</div>
`;

exports[`should render total for two expenses and pluralize correctly 1`] = `
<div>
  <p>
    Viewing 
    2
     
    expenses
     totalling 
    $851.95
  </p>
</div>
`;
